### ISSUE TITLE
ugly errors with tracebacks when tls/ssl handshake fails

### ISSUE THREAD

**Issue by wbolster** [benoitc/gunicorn#1574]
clients that fail to complete the tls/ssl handshake successfully result in quite verbose and ugly errors with tracebacks in gunicorn's output:

```
[2017-08-24 10:53:33 +0200] [11741] [ERROR] Exception in worker process:
Traceback (most recent call last):
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 129, in handle
    req = six.next(parser)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/parser.py", line 41, in __next__
    self.mesg = self.mesg_class(self.cfg, self.unreader, self.req_count)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/message.py", line 153, in __init__
    super(Request, self).__init__(cfg, unreader)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/message.py", line 53, in __init__
    unused = self.parse(self.unreader)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/message.py", line 165, in parse
    self.get_data(unreader, buf, stop=True)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/message.py", line 156, in get_data
    data = unreader.read()
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/unreader.py", line 38, in read
    d = self.chunk()
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/unreader.py", line 65, in chunk
    return self.sock.recv(self.mxchunk)
  File "/usr/lib/python3.6/ssl.py", line 987, in recv
    return self.read(buflen)
  File "/usr/lib/python3.6/ssl.py", line 865, in read
    return self._sslobj.read(len, buffer)
  File "/usr/lib/python3.6/ssl.py", line 627, in read
    v = self._sslobj.read(len)
ssl.SSLError: [SSL: SSL_HANDSHAKE_FAILURE] ssl handshake failure (_ssl.c:2171)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/arbiter.py", line 515, in spawn_worker
    worker.init_process()
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/base.py", line 126, in init_process
    self.run()
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 119, in run
    self.run_for_one(timeout)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 66, in run_for_one
    self.accept(listener)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 30, in accept
    self.handle(listener, client, addr)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 141, in handle
    self.handle_error(req, client, addr, e)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/base.py", line 213, in handle_error
    self.log.exception("Error handling request %s", req.uri)
AttributeError: 'NoneType' object has no attribute 'uri'
Traceback (most recent call last):
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 129, in handle
    req = six.next(parser)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/parser.py", line 41, in __next__
    self.mesg = self.mesg_class(self.cfg, self.unreader, self.req_count)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/message.py", line 153, in __init__
    super(Request, self).__init__(cfg, unreader)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/message.py", line 53, in __init__
    unused = self.parse(self.unreader)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/message.py", line 165, in parse
    self.get_data(unreader, buf, stop=True)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/message.py", line 156, in get_data
    data = unreader.read()
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/unreader.py", line 38, in read
    d = self.chunk()
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/http/unreader.py", line 65, in chunk
    return self.sock.recv(self.mxchunk)
  File "/usr/lib/python3.6/ssl.py", line 987, in recv
    return self.read(buflen)
  File "/usr/lib/python3.6/ssl.py", line 865, in read
    return self._sslobj.read(len, buffer)
  File "/usr/lib/python3.6/ssl.py", line 627, in read
    v = self._sslobj.read(len)
ssl.SSLError: [SSL: SSL_HANDSHAKE_FAILURE] ssl handshake failure (_ssl.c:2171)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/arbiter.py", line 515, in spawn_worker
    worker.init_process()
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/base.py", line 126, in init_process
    self.run()
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 119, in run
    self.run_for_one(timeout)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 66, in run_for_one
    self.accept(listener)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 30, in accept
    self.handle(listener, client, addr)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/sync.py", line 141, in handle
    self.handle_error(req, client, addr, e)
  File "/home/wbolster/Projects/foobar/.direnv/python-3.6.2/lib/python3.6/site-packages/gunicorn/workers/base.py", line 213, in handle_error
    self.log.exception("Error handling request %s", req.uri)
AttributeError: 'NoneType' object has no attribute 'uri'
[2017-08-24 10:53:33 +0200] [11741] [INFO] Worker exiting (pid: 11741)
[2017-08-24 10:53:34 +0200] [11807] [INFO] Booting worker with pid: 11807
```

the

```
AttributeError: 'NoneType' object has no attribute 'uri'
```

near the end seems to indicate that `handle_error()` actually raises an exception itself.

**Comment by toddpalino** [comment]
I've been seeing something similar to this as well, but in the gthread worker.

```2017-09-14 13:10:52 [10995] [ERROR] Exception in worker process
Traceback (most recent call last):
  File "/export/content/lid/apps/menagerie/67ee80cfe8f561cc2d0df89b280a133b0e4d6fb5/libexec/install/gunicorn-19.6.0-py2.py3-none-any.whl.d10768d80eea18ae482418f6a7b97e4ed7ee298f/gunicorn-19.6.0-py2.py3-none-any.whl/gunicorn/arbiter.py", line 557, in spawn_worker
    worker.init_process()
  File "/export/content/lid/apps/menagerie/67ee80cfe8f561cc2d0df89b280a133b0e4d6fb5/libexec/install/gunicorn-19.6.0-py2.py3-none-any.whl.d10768d80eea18ae482418f6a7b97e4ed7ee298f/gunicorn-19.6.0-py2.py3-none-any.whl/gunicorn/workers/gthread.py", line 109, in init_process
    super(ThreadWorker, self).init_process()
  File "/export/content/lid/apps/menagerie/67ee80cfe8f561cc2d0df89b280a133b0e4d6fb5/libexec/install/gunicorn-19.6.0-py2.py3-none-any.whl.d10768d80eea18ae482418f6a7b97e4ed7ee298f/gunicorn-19.6.0-py2.py3-none-any.whl/gunicorn/workers/base.py", line 132, in init_process
    self.run()
  File "/export/content/lid/apps/menagerie/67ee80cfe8f561cc2d0df89b280a133b0e4d6fb5/libexec/install/gunicorn-19.6.0-py2.py3-none-any.whl.d10768d80eea18ae482418f6a7b97e4ed7ee298f/gunicorn-19.6.0-py2.py3-none-any.whl/gunicorn/workers/gthread.py", line 216, in run
    callback(key.fileobj)
  File "/export/content/lid/apps/menagerie/67ee80cfe8f561cc2d0df89b280a133b0e4d6fb5/libexec/install/gunicorn-19.6.0-py2.py3-none-any.whl.d10768d80eea18ae482418f6a7b97e4ed7ee298f/gunicorn-19.6.0-py2.py3-none-any.whl/gunicorn/workers/gthread.py", line 137, in accept
    self.enqueue_req(conn)
  File "/export/content/lid/apps/menagerie/67ee80cfe8f561cc2d0df89b280a133b0e4d6fb5/libexec/install/gunicorn-19.6.0-py2.py3-none-any.whl.d10768d80eea18ae482418f6a7b97e4ed7ee298f/gunicorn-19.6.0-py2.py3-none-any.whl/gunicorn/workers/gthread.py", line 125, in enqueue_req
    conn.init()
  File "/export/content/lid/apps/menagerie/67ee80cfe8f561cc2d0df89b280a133b0e4d6fb5/libexec/install/gunicorn-19.6.0-py2.py3-none-any.whl.d10768d80eea18ae482418f6a7b97e4ed7ee298f/gunicorn-19.6.0-py2.py3-none-any.whl/gunicorn/workers/gthread.py", line 65, in init
    **self.cfg.ssl_options)
  File "/export/apps/python/3.5/lib/python3.5/ssl.py", line 1077, in wrap_socket
    ciphers=ciphers)
  File "/export/apps/python/3.5/lib/python3.5/ssl.py", line 760, in __init__
    self.do_handshake()
  File "/export/apps/python/3.5/lib/python3.5/ssl.py", line 996, in do_handshake
    self._sslobj.do_handshake()
  File "/export/apps/python/3.5/lib/python3.5/ssl.py", line 641, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLEOFError: EOF occurred in violation of protocol (_ssl.c:720)
2017-09-14 13:10:52 [10995] [INFO] Worker exiting (pid: 10995)
```

It looks like a general failure to catch and handle SSL errors during the handshake. It's not only an ugly error, it causes the worker process to die and get restarted for every exception.

**Comment by wbolster** [comment]
this might actually a duplicate of the long fixed #1177. i will close this for now.
